cranelift/zkasm_data/benchmarks/fibonacci/generated/handwritten.zkasm

```diff
@@ -0,0 +1,28 @@
+start:
+  10000 => A
+  :CALL(fib)
+fibEnd:
+  15574651946073070043n :ASSERT
+finalizeExecution:
+  ${beforeLast()}  :JMPN(finalizeExecution)
+                   :JMP(start)
+
+; Computes the fibonacci number for N stored in register A.
+; Stores the result in register A.
+; Uses registers: C, D.
+fib:
+  A :JMPZ(fibReturn)
+  ; Here A >= 1.
+
+  A => D
+  0 => A
+  1 => B
+fibLoop:
```

Comment:
FYI, the fastest way to compute fibo numbers is ``O(log2(n))``. 
The formula: ``F(n)=(\phi^n - (1 - \phi)^n) / sqrt(5)``, where ``\phi`` is golden ratio, or ``(1+sqrt(5))/2``. I played a bit with godbolt, but didn't find a case, where any compiler detect fibo loop and replace by this formula, so, maybe, loop is really best case.

Reply:
And, I was a bit surprised that llvm don't detect such loops, it usually replace loops for arithmetic progressions, I was guessing it is just manually ``if`` for some specific loops. Maybe good place for first commit to llvm (=

Reply:
Well, if we go down that rabbit hole, an even faster way (on modern processors) would be to use matrix exponentiation https://codeforces.com/blog/entry/67776 ;)

But yes, I wouldn't expect a compiler to apply such optimization as it is quite a niche. Completely coincidentally there is a "compiler" that supports this written by NEAR co-founder https://kukuruku.co/hub/algorithms/using-the-quick-raise-of-matrices-to-power-to-write-a-very-fast-interpreter-of-a-simple-programming-language